crypto/bio/bss_mem.c

```diff
@@ -296,13 +314,18 @@ static long mem_ctrl(BIO *b, int cmd, long num, void *ptr)
         if (ptr != NULL) {
             pptr = (char **)ptr;
             *pptr = (char *)(bm->data);
+            b->flags |= BIO_FLAGS_DATA_FETCH;
```

Comment:
I am not convinced we should change the behaviour of `BIO_CTRL_INFO` (BIO_get_mem_data). It is just to get the buffer and size for informational purposes. We don't document this function as changing the ownership semantics of the buffer; it's the application's responsibility to ensure it doesn't use the buffer beyond the lifetime as it remains managed by the BIO.

Reply:
For 3.1 I could agree with that, I'm really just trying to find a way (thats API compatible) to allow users a method to orphan the data pointer without orphaning and leaking the BUF_MEM object.  Currently the only available leak free way to do that is 
`
BIO_get_mem_data(bio, &ptr);
BIO_get_mem_ptr(bio, &bmptr);
bmptr->data = NULL;
BIO_free(bio);
`
which isn't super clear.  Perhaps its better to just update the documentation to explicitly call this out in 3.1 and find a better (allbeit API incompatible) way to do this in a future release

FWIW, my plan for the master branch is to propose removing BIO_get_mem_data entirely and replace it with a BUF_MEM_free_data(bio, &ptr) method that orphans the data at free time, that would result in the following code which I think is a bit more clear:
`
BIO_get_mem_ptr(bio, &bmptr);
BIO_set_close(bio, BIO_NOCLOSE);
BIO_free(bio);
BUF_MEM_free(bmptr, &ptr);
`

Reply:
Updating the docs for 3.1 is definitely a good idea. I also agree it would be valuable to provide this capability. Perhaps a new ctrl?